### PR TITLE
Adding IsDisposed checks to Span getter and GetArray/Pointer Internal

### PR DESCRIPTION
**Pull request by ahsonkhan**
cc @KrzysztofCwalina, @pakrym, @davidfowl

Addressing comments:
https://github.com/dotnet/corefxlab/pull/1465#discussion_r112050277
https://github.com/dotnet/corefxlab/pull/1465#discussion_r112050352

### CODE REVIEW

File: samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs [hunk 1]
```diff
@@ -120,12 +127,12 @@ protected override void Dispose(bool disposing)
 
         public override Span<byte> GetSpan(int index, int length)
```
Comment (pakrym) [inline samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs]: Why do we need this method? What is the case where it's implementation is different from slicing `.Span`?

File: samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs [hunk 2]
```diff
@@ -120,12 +127,12 @@ protected override void Dispose(bool disposing)
...
         {
-            if (IsDisposed) ThrowObjectDisposed();
             return Span.Slice(index, length);
         }
 
         protected override bool TryGetArrayInternal(out ArraySegment<byte> buffer)
         {
+            if (IsDisposed) ThrowObjectDisposed();
```
Comment (pakrym) [inline samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs]: Having `ThrowObjectDisposed` and `ThrowIndexOutOfRange` on a public abstract class is bad, use `ThrowHelper`

File: tests/System.Buffers.Primitives.Tests/MemoryTests.cs [hunk 3]
```diff
@@ -71,6 +71,24 @@ static void TestLifetime(OwnedBuffer<byte> owned)
         }
 
 
+        [Fact]
+        public void TestThrowOnAccessAfterDipose()
+        {
+            var array = new byte[1024];
+            OwnedBuffer<byte> owned = array;
```
Comment (pakrym) [inline tests/System.Buffers.Primitives.Tests/MemoryTests.cs]: What about all other `Owned*` implemenations?

Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/MemoryTests.cs]: Many are in samples or experimental. I will add tests for other implementations in System.Buffer.Primitives.

File: samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs (hunk 1 above)
Comment (ahsonkhan) [inline samples/LowAllocationWebServer/LowAllocationWebServerLibrary/OwnedBuffer.cs]: Yes, we can remove the method.

File: src/System.IO.Pipelines/MemoryPoolBlock.cs [hunk 4]
```diff
@@ -39,7 +39,14 @@ protected MemoryPoolBlock(MemoryPool pool, MemoryPoolSlab slab, int offset, int
 
         public override int Length => _length;
 
-        public override Span<byte> Span => new Span<byte>(Slab.Array, _offset, _length);
+        public override Span<byte> Span
+        {
+            get
+            {
+                if (IsDisposed) ThrowHelper.ThrowObjectDisposedException(nameof(MemoryPoolBlock));
+                return new Span<byte>(Slab.Array, _offset, _length);
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/MemoryPoolBlock.cs]: Is there a perfromance difference between the Span ctor that takes array vs the one that takes a pointer? Might be worth filing and issue to find it out, i.e. don't block this PR on that investigation.

File: src/System.IO.Pipelines/ThrowHelper.cs [hunk 5]
```diff
@@ -52,6 +52,11 @@ public static void ThrowArgumentOutOfRangeException_BufferRequestTooLarge(int ma
             throw GetArgumentOutOfRangeException_BufferRequestTooLarge(maxSize);
         }
 
+        public static void ThrowObjectDisposedException(string objectName)
+        {
+            throw GetObjectDisposedException(objectName);
+        }
+
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/ThrowHelper.cs]: BTW, the reason I had these methods on OwnedBuffer (as opposed on throw helper) is that the ThrowHelper is no public, and so each assembly that implements OwnedBuffer needs its own ThrowHelper. @pakrym.

**Comment by KrzysztofCwalina** [comment]
in general looks good.

File: src/System.IO.Pipelines/ThrowHelper.cs (hunk 5 above)
Comment (benaadams) [inline src/System.IO.Pipelines/ThrowHelper.cs]: Trouble is OwnedBuffer is generic `<T>` which means all T's get there own ThrowHelper as well

Comment (pakrym) [inline src/System.IO.Pipelines/ThrowHelper.cs]: And having throw helpers on public API is no good

**Comment by pakrym** [comment]
@KrzysztofCwalina  We need to figure our OwnedBuffer recycling and fix dispose implementation therer.

File: src/System.IO.Pipelines/ThrowHelper.cs (hunk 5 above)
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/ThrowHelper.cs]: Yeah, good point about the cost of generic instantiations

**Pull request merged**